src/ariel-os-coap/src/transport_slipmux.rs

```diff
@@ -0,0 +1,337 @@
+
+    //let uart = usb_serial::take().await;
+    let mut uart = over_uart::take().await;
+
+    let mut decoder = slipmux::Decoder::new();
+    // Small -- just enough so we can show some bytes for diagnostics. (We can do larger if
+    // ReferencedLatestFrame learns to use a single buffer).
+    let mut diag_buf = [0; 8];
+    // Could do less, but don't want to do the math right now.
+    let mut coap_buf = [0; 1200];
+    // No point in having anything while we don't forward.
+    // (Or should we have a few bytes just so we can respond with ICMP No Sorry?)
+    let mut packet_buf = [0; 0];
+    let mut slipmux =
+        slipmux::ReferencedLatestFrame::new(&mut diag_buf, &mut coap_buf, &mut packet_buf);
+    loop {
+        // FIXME: This is an ugly reading mechanism, made just slightly more tolerable by
+        // decoder.decode() taking things bytewise anyway.
+        //
+        // Once <https://github.com/ariel-os/ariel-os/pull/1600> is resolved, this paragraph should
+        // be replaced by the `//BR` comments, which avoid going back and forth with the
+        // peripheral.
+        let mut bytebuf = [0];
+        let read = uart.read(&mut bytebuf).await.unwrap();
+        assert_eq!(read, 1);
+        let byte = bytebuf[0];
+
+        //BR let mut buffer = uart.fill_buf().await.unwrap();
+        //BR let mut consume = 0;
+
+        let mut coap_ready = false;
+
+        //BR while let Some(&byte) = buffer.split_off_first() {
+        //BR consume += 1;
+
+        // Ignoring the return value; handling everything inside the decoder.
+        match decoder.decode(byte, &mut slipmux) {
+            Err(e) => {
+                warn!("Decoding error; trying at the next byte.");
+                //BR consume = 1;
+                //BR break;
+            }
+            Ok(DecodeStatus::Incomplete) => {
+                //BR continue
+            }
+            Ok(DecodeStatus::FrameCompleteDiagnostic) => {
+                // Use up to the cursor, and silently ignore overflows.
+                let buffer = slipmux
+                    .diagnostic_buffer
+                    .get(0..slipmux.index)
+                    .unwrap_or(slipmux.diagnostic_buffer);
+                let text = core::str::from_utf8(buffer);
+                warn!(
+                    "Peer sent diagnostic data. This will no be forwarded; content was {:?}",
+                    text.map_err(|e| &buffer)
+                );
+                //BR break;
+            }
+            Ok(DecodeStatus::FrameCompleteIp) => {
+                warn!("Peer sent network data, which is unsupported.",);
+                //BR break;
+            }
+            Ok(DecodeStatus::FrameCompleteConfiguration) => {
+                coap_ready = true;
+                //BR break;
+            }
+        }
+
+        //BR }
+
+        //BR uart.consume(consume);
+
+        if coap_ready {
+            // Compensate for checksum -- FIXME: where should this best be done?
```

Comment:
If you come up with a good idea, let me know! I would prefer for the `slipmux` crate to handel this but it is not as straight forward as one thinks: You can only know that you deserialezed the fcs after the frame completed, at which point the bytes have already been forwarded to the handler which, well, is in the users hands. And here we are >.<